core/drivers/clk/sam/at91_programmable.c

```diff
@@ -0,0 +1,163 @@
+					       unsigned long parent_rate)
+{
+	struct clk_programmable *prog = clk->priv;
+	const struct clk_programmable_layout *layout = prog->layout;
+	unsigned int pckr = io_read32(prog->base + AT91_PMC_PCKR(prog->id));
+	unsigned long rate = 0;
+
+	if (layout->is_pres_direct)
+		rate = parent_rate / (PROG_PRES(layout, pckr) + 1);
+	else
+		rate = parent_rate >> PROG_PRES(layout, pckr);
+
+	return rate;
+}
+
+static TEE_Result clk_programmable_set_parent(struct clk *clk, size_t index)
+{
+	struct clk_programmable *prog = clk->priv;
+	const struct clk_programmable_layout *layout = prog->layout;
+	unsigned int mask = layout->css_mask;
+	unsigned int pckr = index;
+
+	if (layout->have_slck_mck)
+		mask |= AT91_PMC_CSSMCK_MCK;
+
+	if (index > layout->css_mask) {
+		if (index > PROG_MAX_RM9200_CSS && !layout->have_slck_mck)
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		pckr |= AT91_PMC_CSSMCK_MCK;
+	}
+
+	io_clrsetbits32(prog->base + AT91_PMC_PCKR(prog->id), mask, pckr);
+
+	return TEE_SUCCESS;
+}
+
+static size_t clk_programmable_get_parent(struct clk *clk)
+{
+	struct clk_programmable *prog = clk->priv;
+	const struct clk_programmable_layout *layout = prog->layout;
+	unsigned int pckr = io_read32(prog->base + AT91_PMC_PCKR(prog->id));
+	size_t ret = 0;
+
+	ret = pckr & layout->css_mask;
+
+	if (layout->have_slck_mck && (pckr & AT91_PMC_CSSMCK_MCK) && !ret)
+		ret = PROG_MAX_RM9200_CSS + 1;
+
+	return ret;
+}
+
+static unsigned int flsi(unsigned int val)
+{
+	if (val == 0)
+		return 0;
+
+	return sizeof(unsigned int) * 8 - __builtin_clz(val);
+}
+
+static TEE_Result clk_programmable_set_rate(struct clk *clk, unsigned long rate,
+					    unsigned long parent_rate)
+{
+	struct clk_programmable *prog = clk->priv;
+	const struct clk_programmable_layout *layout = prog->layout;
+	unsigned long div = parent_rate / rate;
+	int shift = 0;
+
+	if (!div)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (layout->is_pres_direct) {
+		shift = div - 1;
+
+		if (shift > layout->pres_mask)
+			return TEE_ERROR_BAD_PARAMETERS;
+	} else {
+		shift = flsi(div) - 1;
+
+		if (div != (1ULL << shift))
+			return TEE_ERROR_BAD_PARAMETERS;
+
+		if (shift >= layout->pres_mask)
+			return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	io_clrsetbits32(prog->base + AT91_PMC_PCKR(prog->id),
+			layout->pres_mask << layout->pres_shift,
+			shift << layout->pres_shift);
+
+	return TEE_SUCCESS;
+}
+
+static const struct clk_ops programmable_ops = {
+	.get_rate = clk_programmable_get_rate,
+	.get_parent = clk_programmable_get_parent,
+	.set_parent = clk_programmable_set_parent,
+	.set_rate = clk_programmable_set_rate,
+};
+
+struct clk *
+at91_clk_register_programmable(struct pmc_data *pmc,
+			       const char *name, struct clk **parents,
+			       uint8_t num_parents, uint8_t id,
+			       const struct clk_programmable_layout *layout)
+{
+	struct clk_programmable *prog = NULL;
+	struct clk *clk = NULL;
+
+	if (id > PROG_ID_MAX)
+		return NULL;
```

Comment:
should not happen, right?  Assert or panic?

Reply:
Should never happen, adding an `assert()`.